Add map folding tutorial (#824)
dhall-lang/dhall-lang: docs/tutorials/string-matrix.md

# Using map and fold to generate a list of strings

The goal of this tutorial is to generate a YAML list of tests to run, such as:

```yaml
- test-variantA-suffixA
- test-variantA-suffixB
- test-variantB-suffixA
- test-variantB-suffixB
```

We'll use the `dhall-to-yaml` command line, make sure to follow the [install instructions][installation]
from the getting started tutorial.

[installation]: <tutorials/Getting-started_Generate-JSON-or-YAML:installation>

## Simple map

[Map](https://en.wikipedia.org/wiki/Map_%28higher-order_function%29) is a function
that applies a function to each element of a list, for example:

```dhall
-- ./string-matrix.dhall
let List/map =
      https://prelude.dhall-lang.org/v11.1.0/List/map sha256:dd845ffb4568d40327f2a817eb42d1c6138b929ca758d50bc33112ef3c885680

let generate-test-name =
          \(prefix : Text)
      ->  List/map Text Text (\(suffix : Text) -> "${prefix}-${suffix}")

in  generate-test-name "variantA" [ "suffixA", "suffixB" ]
```
```console

> $ dhall-to-yaml --file ./string-matrix.dhall

Reviewer: I would also include the output of the `dhall --file` command, too, for users less familiar with YAML:

```bash
$ dhall --file ./string-matrix.dhall
[ "variantA-suffixA", "variantA-suffixB" ]
```

- variantA-suffixA
- variantA-suffixB
```

> In dhall, generic functions such as `map` require abstract types to be explicit. The `map` signature is:

Reviewer: ```suggestion
In dhall, generic functions such as `map` require types to be explicitly specified as function arguments. The `map` signature is:
```

> * input type (A),

Reviewer: I would include the full type signature for `map` here before explaining the components of the type

* output type (B),
* a function that goes from A to B,
* a list of A, and
* return a list of B.

> Functions can be pre-applied as each argument evaluates to a new function.

Reviewer: ```suggestion
Functions can be partially saturated as each argument evaluates to a new function.
```

This technique is called [currying](https://en.wikipedia.org/wiki/Currying).

> For example, the above `List/map` call successively evaluates as:

Reviewer: ```suggestion
For example, the type of the `List/map` function changes with each new argument:
```

```dhall
List/map
    : forall (a : Type) -> forall (b : Type) -> (a -> b) -> List a -> List b

List/map Text
    : forall (b : Type) -> (Text -> b) -> List Text -> List b

List/map Text Text
    : (Text -> Text) -> List Text -> List Text

List/map Text Text (\(prefix : Text) -> "${prefix}-${suffix}")
    : List Text -> List Text
```

Thus, `generate-test-name` pre-applies the `map` function from `Text` to `Text` using a known prefix.
Then it is given a list of suffixes to generate new strings.

## Double map

> We can call two `map`s and generate our initial goal by adding such a function to our `string-matrix.dhall` file:

Reviewer: ```suggestion
We can call two `map`s and generate our initial goal by adding the following function to our `./string-matrix.dhall` file:
```

```dhall
let generate-test-suite =
          \(prefixes : List Text)
      ->  \(suffixes : List Text)
      ->  List/map
            Text
            (List Text)
            (\(prefix : Text) -> generate-test-name prefix suffixes)
            prefixes

in  generate-test-suite [ "variantA", "variantB" ] [ "suffixA", "suffixB" ]
```
```console

> $ dhall-to-yaml --file ./string-matrix.dhall

Reviewer: Same suggestion here: also include the output of the `dhall` command for reference

- - variantA-suffixA
  - variantA-suffixB
- - variantB-suffixA
  - variantB-suffixB
```

However, by doing so we get a value of type `List (List Text)` (i.e. a `List` of `List`s of `Text`).
So we can get a flat list by using the fold function.

## Folding a list of list of texts to a flat list of text

> [Fold](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) is a function

Reviewer: ```suggestion
A ["fold"](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) is a function
```

that reduces a list of objects to another type. Its signature is:

> * input type (A),

Reviewer: Same suggestion here: include the full type signature for `List/fold` first before explaining the components of the type

* the list to fold (a List of A),
* output type (B),
* a folding function that goes from (A, B) to B,
* a starting value of type B, and
* return a B.

> Therefore, we can implement a `fold-list-text` function like so:

Reviewer: Perhaps call this function "`flatten-list-text`" since once you saturate the `List/fold` function the result is no longer a "fold"

Author: Good point.

in  generate-test-suite [ "variantA", "variantB" ] [ "suffixA", "suffixB" ]
```

This is equivalent to the previous evaluation, but the fold-list-text function is now hidden in
the generate-test-suite thanks to a nested `let`.

We could use the new function to generate even more variations:

```dhall
in  generate-test-suite
      ( generate-test-suite
          (generate-test-suite [ "tox" ] [ "py27", "py38" ])
          [ "postgresql", "sqlite" ]
      )
      [ "el6", "el7" ]
```
```yaml
- tox-py27-postgresql-el6
- tox-py27-postgresql-el7
- tox-py27-sqlite-el6
- tox-py27-sqlite-el7
- tox-py38-postgresql-el6
- tox-py38-postgresql-el7
- tox-py38-sqlite-el6
- tox-py38-sqlite-el7
```

## Generalizing the string matrix generation

While the above is working, the code can be improved to remove repetitive function calls.
The [fold](https://prelude.dhall-lang.org/List/fold) documentation actually mentions:

> If you treat the list `[ x, y, z ]` as `cons x (cons y (cons z nil))`, then a
> `fold` just replaces each `cons` and `nil` with something else.

This is exactly what we need, `cons` is our `generate-test-suite` and `nil` is
the initial suffix `[ "el6", "el7" ]`!

```dhall

> let generate-test-matrix =

Reviewer: It seems like it would be more elegant if the `generate-test-matrix` function had the following type:

```dhall
generate-test-matrix : List (List Text) -> List Text
```

... so that you could then write:

```dhall
generate-test-matrix [ [ "tox" ], ["py27", "py38" ], [ "postgresql", "sqlite" ], [ "el6", "el7" ] ]
```

Would it be possible to extend the tutorial to show how to do that?

Author: It seems like we either have to use `List/last` (and introduce optionals), or rewrite the `generate-test-suite` function to handle empty initial prefixes. I meant to implement the first solution in another tutorial doc to keep it simple, but this can be added to this doc instead.

          \(variations : List (List Text))
      ->  \(last : List Text)
      ->  List/fold (List Text) variations (List Text) generate-test-suite last

in  generate-test-matrix
      [ [ "tox" ], [ "py27", "py38" ], [ "postgresql", "sqlite" ] ]
      [ "el6", "el7" ]
```
```yaml
- tox-py27-postgresql-el6
- tox-py38-postgresql-el6
- tox-py27-sqlite-el6
- tox-py38-sqlite-el6
- tox-py27-postgresql-el7
- tox-py38-postgresql-el7
- tox-py27-sqlite-el7
- tox-py38-sqlite-el7
```

> **Note**: Using the `List/head` function the `generate-test-matrix` could
> be improved to take a single list of argument.

> > A follow-up tutorial introduces that using `Optional/fold`

Reviewer: Is this note up-to-date?  I couldn't find the `List/head` that this was referring to

Author: Yeah, the improvement still needs to be written... Perhaps from a tutorial point of view we could explain why folding with `generate-test-suite` needs a starting prefix and that `generate-test-matrix` could be improved by using `List/last` as an exercise left to the reader. But I found `Optional/fold` to be unintuitive to use and figured this should be introduced in its own tutorial.

I meant to cover a couple more topics such as using List/filter to blacklist some combination (e.g. skip py3 on el6) and rendering more complex structure such as
```yaml
- job:
    name: tox-py38-sqlite
    vars:
      version: py36
      backend: sqlite
```
I think that would make a nice 4 parts tutorial.

# Using map and fold to generate a list of strings

The goal of this tutorial is to generate a YAML list of test names, such as:

```yaml
- test-variantA-suffixA
- test-variantA-suffixB
- test-variantB-suffixA
- test-variantB-suffixB
```

We'll use the `dhall` and `dhall-to-yaml` command line,
make sure to follow the [install instructions][installation] from the getting started tutorial.

[installation]: <tutorials/Getting-started_Generate-JSON-or-YAML:installation>

> ## Simple map

Reviewer: ```suggestion
## Map
```

I try to avoid the use of the word "simple" in tutorial material, since not all users might consider the concepts introduced within a tutorial to be simple

Author: Make sense

> A ["map"](https://en.wikipedia.org/wiki/Map_%28higher-order_function%29) is a function

Reviewer: I think this should be preceded by a paragraph motivating why we want to apply a function to each element of a list.  The reason to do so is that you want to put the reader in a frame of mind where "If I have a goal X then the solution is Y" so that the next time they have the same goal X they immediately remember to use solution Y.

Currently the prose jumps straight to the solution Y (e.g. "use a combinator that applies a function to each element of the list") without explaining why we need such a function.  Here's an example of how I might motivate the use of `map`:

> We would like to begin by prepending the same prefix to a list of suffixes, so that our final code looks like this:
>
> ```dhall
> generate-test-name "variantA" [ "suffixA", "suffixB" ]
> 
> = [ "variantA-suffixA", "variantA-suffixB" ]
> ```
>
> ... but in order to do so we need a way to apply the same transformation (e.g. append a suffix) to each element of a list.
>
> The ["map"](…) higher-order function provides precisely this capability: we can use `map` to uniformly transform each element of a list, like this:
>
> …

I have a similar comment for the following two sections: first explain to the user what the local goal is for that section before introducing the utility or design pattern that realizes that goal.

Author: Thank you for the suggestion. By the way, I'm really digging the tutorial/how-to/discussions/references doc organization after going through dhall doc :)

that applies a function to each element of a list, for example:

```dhall
-- ./string-matrix.dhall

let List/map =
      https://prelude.dhall-lang.org/v11.1.0/List/map sha256:dd845ffb4568d40327f2a817eb42d1c6138b929ca758d50bc33112ef3c885680

let generate-test-name =
          \(prefix : Text)
      ->  List/map Text Text (\(suffix : Text) -> "${prefix}-${suffix}")

in  generate-test-name "variantA" [ "suffixA", "suffixB" ]
```
```console
$ dhall --file ./string-matrix.dhall
[ "variantA-suffixA", "variantA-suffixB" ]

$ dhall-to-yaml --file ./string-matrix.dhall
- variantA-suffixA
- variantA-suffixB
```

In dhall, generic functions such as `map` require types to be explicitly specified as function arguments. The `map` signature is:

```dhall
let map : forall (a : Type) -> forall (b : Type) -> (a -> b) -> List a -> List b
```

In other words, the `map` arguments are:

> * input type (A),

Reviewer: ```suggestion
* input type (`a`),
```

For consistency with the type signature, I suggest using lowercase when referencing the types (and format the symbols as code)

* output type (B),
* a function that goes from A to B,
* a list of A, and
* it returns a list of B.

Functions can be partially saturated as each argument evaluates to a new function.
This technique is called [currying](https://en.wikipedia.org/wiki/Currying).
For example, the type of the `List/map` function changes with each new argument:

```dhall
List/map
    : forall (a : Type) -> forall (b : Type) -> (a -> b) -> List a -> List b

List/map Text
    : forall (b : Type) -> (Text -> b) -> List Text -> List b

List/map Text Text
    : (Text -> Text) -> List Text -> List Text

List/map Text Text (\(prefix : Text) -> "${prefix}-${suffix}")
    : List Text -> List Text
```

Thus, `generate-test-name` pre-applies the `map` function from `Text` to `Text` using a known prefix.
Then it is given a list of suffixes to generate new strings.

> ## Double map

Reviewer: ```suggestion
## Nested map
```

This is a minor suggestion to indicate that the `map`s are not being applied in succession (e.g. `map f (map g xs)`) but are instead being applied in a manner analogous to a nested `for` loop

We can call two `map`s and generate our initial goal by adding the following function to our `./string-matrix.dhall` file:

```dhall

> let generate-test-suite =

Reviewer: I suggest including the entire program at each step so that the user can follow along at each stage by copying and pasting the entire example

The other reason I suggest this is because the functions have similar names (e.g. `generate-*`) so including both of them side-by-side visually indicates that they are now working with two distinct functions instead of amending the old `generate-test-name` function)

Author: That's a good idea.

in  generate-test-suite [ "variantA", "variantB" ] [ "suffixA", "suffixB" ]
```
```console
$ dhall --file ./string-matrix.dhall
[ [ "variantA-suffixA", "variantA-suffixB" ]
, [ "variantB-suffixA", "variantB-suffixB" ]
]

$ dhall-to-yaml --file ./string-matrix.dhall
- - variantA-suffixA
  - variantA-suffixB
- - variantB-suffixA
  - variantB-suffixB
```

However, by doing so we get a value of type `List (List Text)` (i.e. a `List` of `List`s of `Text`).
So we can get a flat list by using the fold function.

## Folding a list of list of texts to a flat list of text

A ["fold"](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) is a function
that reduces a list of objects to another type. Its signature is:

```dhall
let fold
    :   forall (a : Type)
      → List a
      → forall (list : Type)
      → forall (cons : a → list → list)
      → forall (nil : list)
      → list
```

In other words, the `fold` arguments are:

* input type (A),
* the list to fold (a List of A),
* output type (B),
* a folding function that goes from (A, B) to B,
* a starting value of type B, and
* it returns a B.

Therefore, we can implement a `flatten-list-text` function like so:

```dhall

> -- concat-example.dhall

Reviewer: ```suggestion
-- ./concat-example.dhall
```

> let {- Convert [[a, b], [c, d]] to [a, b, c, d] -}

Reviewer: You can actually turn this comment into an `assert` test! 🙂 

```dhall
let flatten-list-text =
        λ(to-fold : List (List Text))
      →   List/fold
            (List Text)
            to-fold
            (List Text)
            (λ(a : List Text) → λ(b : List Text) → a # b)
            ([] : List Text)
        : List Text

let example =
        λ(a : Text)
      → λ(b : Text)
      → λ(c : Text)
      → λ(d : Text)
      → assert : flatten-list-text [ [ a, b ], [ c, d ] ] ≡ [ a, b, c, d ]

in  flatten-list-text [ [ "A", "B" ], [ "C", "D" ] ]
```

Author: That's really neat how such a test can be asserted. Thanks for the example.

    flatten-list-text =
          \(to-fold : List (List Text))
      ->    List/fold
              (List Text)
              to-fold
              (List Text)
              (\(a : List Text) -> \(b : List Text) -> a # b)
              ([] : List Text)
          : List Text

in  flatten-list-text [ [ "A", "B" ], [ "C", "D" ] ]
```

```console
$ dhall-to-yaml --file ./concat-example.dhall

> [A, B, C, D]

Reviewer: Is this output correct?  I would have expected:

```yaml
- A
- B
- C
- D
```

Author: That was not correct indeed.

```

> **NOTE**: This function can also be defined as:
> ```dhall
> let flatten-list-text = https://prelude.dhall-lang.org/List/concat Text
>

We can use this new function in our previous example:
```dhall
in flatten-list-text
     (generate-test-suite [ "suffixA", "suffixB" ] [ "variantA", "variantB" ])
```
```yaml
- variantA-suffixA
- variantA-suffixB
- variantB-suffixA
- variantB-suffixB
```

## Refactoring the generate-test-suite function

> The flatten-list-text function doesn't have to be visible and the generate test-suite

Reviewer: ```suggestion
The flatten-list-text function doesn't have to be a top-level function and the generate test-suite
```
